### PR TITLE
backup: retry on internal error and make more errors can be retried

### LINKED ISSUE #34350: backup should retry on s3 internal error.
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
when s3 return internal error. backup will failed due we didn't handle this kind of error in https://github.com/pingcap/tidb/blob/c0e39e400f1f9e72313d6438289feb2ead83bb9a/br/pkg/backup/client.go#L1032.

perhaps we should retry as much as we can. according to https://grpc.github.io/grpc/core/md_doc_statuscodes.html. I think these following errors can be retried.
```
	// Canceled indicates the operation was canceled (typically by the caller).
	//
	// The gRPC framework will generate this error code when cancellation
	// is requested.
	Canceled Code = 1

	// DeadlineExceeded means operation expired before completion.
	// For operations that change the state of the system, this error may be
	// returned even if the operation has completed successfully. For
	// example, a successful response from a server could have been delayed
	// long enough for the deadline to expire.
	//
	// The gRPC framework will generate this error code when the deadline is
	// exceeded.
	DeadlineExceeded Code = 4

	// ResourceExhausted indicates some resource has been exhausted, perhaps
	// a per-user quota, or perhaps the entire file system is out of space.
	//
	// This error code will be generated by the gRPC framework in
	// out-of-memory and server overload situations, or when a message is
	// larger than the configured maximum size.
	ResourceExhausted Code = 8

	// Aborted indicates the operation was aborted, typically due to a
	// concurrency issue like sequencer check failures, transaction aborts,
	// etc.
	//
	// See litmus test above for deciding between FailedPrecondition,
	// Aborted, and Unavailable.
	//
	// This error code will not be generated by the gRPC framework.
	Aborted Code = 10

	// Internal errors. Means some invariants expected by underlying
	// system has been broken. If you see one of these errors,
	// something is very broken.
	//
	// This error code will be generated by the gRPC framework in several
	// internal error conditions.
	Internal Code = 13

	// Unavailable indicates the service is currently unavailable.
	// This is a most likely a transient condition and may be corrected
	// by retrying with a backoff. Note that it is not always safe to retry
	// non-idempotent operations.
	//
	// See litmus test above for deciding between FailedPrecondition,
	// Aborted, and Unavailable.
	//
	// This error code will be generated by the gRPC framework during
	// abrupt shutdown of a server process or network connection.
	Unavailable Code = 14
```

### 2. What did you expect to see? (Required)
backup succeed.

### 3. What did you see instead (Required)
backup failed.

### 4. What is your TiDB version? (Required)
master

### PR DESCRIPTION
**Pull request by 3pointer**
### What problem does this PR solve?

Issue Number: close #34350 

Problem Summary:
This PR add will try retry as much as we can to improve robustness.

### What is changed and how it works?
make more errors can be retryable( including internal error)
### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Fix the issue that backup cannot handle s3 internal error.
```

### CODE REVIEW

File: br/pkg/backup/client.go [hunk 1]
```diff
@@ -1030,9 +1038,18 @@ const (
...
-	if status.Code(err) == codes.Unavailable {
-		return true
+	// some errors can be retried
+	// https://github.com/pingcap/tidb/issues/34350
+	switch status.Code(err) {
+	case codes.Unavailable: return true
+	case codes.DeadlineExceeded : return true
+	case codes.ResourceExhausted: return true
+	case codes.Aborted: return true
+	case codes.Internal:
```
Comment (hawkingrei) [inline br/pkg/backup/client.go:1048]: Those conditional judgment can be combined together

```
case codes.Unavailable,codes.DeadlineExceeded,...:
```

**Comment by overvenus** [comment]
For errors like

```log
failed to put object Request ID: None Body: <?xml version=\\\"1.0\\\" encoding=\\\"UTF-8\\\"?>\\n<Error><Code>InternalError</Code><Message>We encountered an internal error. Please try again.</Message><RequestId>6M5A92TA8E2JBWHW</RequestId><HostId>Xvx4tukwnPUDlWMK5JbuxshfEfPM4GC52PNwCoamCVcgYtcOShhjzr0xZ5GYcksfwY847jYDKuE=</HostId></Error>\
```

should we add `"InternalError"` to `tidb/br/pkg/utils/retry.retryableServerError` ?

https://github.com/pingcap/tidb/blob/dc5ec235914aeb7c2506d5b5e3eb21493eec2e83/br/pkg/utils/retry.go#L16-L26

**Comment by 3pointer** [comment]
> For errors like
> 
> ```
> failed to put object Request ID: None Body: <?xml version=\\\"1.0\\\" encoding=\\\"UTF-8\\\"?>\\n<Error><Code>InternalError</Code><Message>We encountered an internal error. Please try again.</Message><RequestId>6M5A92TA8E2JBWHW</RequestId><HostId>Xvx4tukwnPUDlWMK5JbuxshfEfPM4GC52PNwCoamCVcgYtcOShhjzr0xZ5GYcksfwY847jYDKuE=</HostId></Error>\
> ```
> 
> should we add `"InternalError"` to `tidb/br/pkg/utils/retry.retryableServerError` ?
> 
> https://github.com/pingcap/tidb/blob/dc5ec235914aeb7c2506d5b5e3eb21493eec2e83/br/pkg/utils/retry.go#L16-L26

I feel like we should add "InternalError" to tidb/br/pkg/utils/retry.retryableServerError here. since tikv has translate internal error to [errorpb](https://github.com/tikv/tikv/blob/v5.0.6/components/backup/src/endpoint.rs#L745). 

but it's too strange that I cannot found related log in ES in newly oncall. maybe ES lost the important logs? 🤔

**Review by YuJuncen — APPROVED** [review]

**Review by fengou1 — APPROVED** [review]

**Review by YuJuncen — APPROVED** [review]

**Pull request merged**